### PR TITLE
Fix debug option in the dlk

### PR DESCRIPTION
**Pull request by nlpng**
## Motivation and Context

This WIP PR fixes the `--debug` option in `generate_project` to properly save the outputs of each layer when running inference.

## Description

[NOTE] Currently this option is only available when the outputs of each layer are in float, if using `-hq`, `-ts` to generate the project, this option may not function properly, since after optimization passes, the output name of operators changed, hence this makes the case to be trickier to retrieve the information properly.

The output should be like the following to identify which layer starts to fail:
```
Checking debug data...
[OK]   conv1_1_pool 0
[OK]   conv1_1_Conv2D 0
[OK]   conv1_1_bn_FusedBatchNorm 0
[OK]   QTZ_linear_mid_tread_half 0
[OK]   quantized_conv1_2_Conv2D 0
[OK]   quantized_conv1_2_bn_FusedBatchNorm 0
[OK]   quantized_QTZ_linear_mid_tread_half 0
[OK]   quantized_conv2_1_pool 0
[OK]   quantized_split 0
[OK]   quantized_conv2_1_0_Conv2D 0
[OK]   quantized_conv2_1_0_bn1_FusedBatchNorm 0
[FAIL] quantized_QTZ_linear_mid_tread_half_1 2
[OK]   quantized_conv2_1_1_Conv2D 0
[OK]   quantized_conv2_1_1_bn1_FusedBatchNorm 0
[OK]   quantized_QTZ_linear_mid_tread_half_2 0
[FAIL] quantized_concat 2
[FAIL] quantized_conv2_1_Conv2D 256
[FAIL] quantized_conv2_1_bn2_FusedBatchNorm 256
[FAIL] quantized_QTZ_linear_mid_tread_half_3 12
[FAIL] quantized_conv3_1_pool 12
[FAIL] quantized_split_1 7
[FAIL] quantized_conv3_1_0_Conv2D 2304
[FAIL] quantized_conv3_1_0_bn1_FusedBatchNorm 2304
[FAIL] quantized_QTZ_linear_mid_tread_half_4 49
[FAIL] quantized_conv3_1_1_Conv2D 2304
[FAIL] quantized_conv3_1_1_bn1_FusedBatchNorm 2304
[FAIL] quantized_QTZ_linear_mid_tread_half_5 50
[FAIL] quantized_concat_1 99
[FAIL] quantized_conv3_1_Conv2D 3860
[FAIL] quantized_conv3_1_bn2_FusedBatchNorm 3860
[FAIL] quantized_QTZ_linear_mid_tread_half_6 239
[FAIL] quantized_split_2 119
[FAIL] quantized_conv5_3_0_Conv2D 5786
[FAIL] quantized_conv5_3_0_bn1_FusedBatchNorm 5786
[FAIL] quantized_QTZ_linear_mid_tread_half_7 274
[FAIL] quantized_conv5_3_1_Conv2D 5621
[FAIL] quantized_conv5_3_1_bn1_FusedBatchNorm 5621
[FAIL] quantized_QTZ_linear_mid_tread_half_8 303
[FAIL] quantized_concat_2 577
[FAIL] quantized_conv5_3_Conv2D 22461
[FAIL] quantized_conv5_3_bn2_FusedBatchNorm 22461
[FAIL] quantized_Relu 13052
[FAIL] gap_AvgPool 112
[FAIL] linear_Reshape 112
[FAIL] linear_MatMul 501
[FAIL] linear_BiasAdd 501
[FAIL] Softmax 104
```

## How has this been tested?

Currently only the networks in float were tested.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: dlk/python/tools/test_debug.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import numpy as np
 import click
 import re
+
```
Comment (tsawada) [inline dlk/python/tools/test_debug.py]: Is this intentional?

File: dlk/python/tools/test_debug.py [hunk 2]
```diff
@@ -53,19 +54,34 @@ def main(debug_data_path, expected_data_path):
...
                 else:
-                    print(f"[FAIL] {name_ex}")
+                    results[ptrn_ex.match(eo.stem).group(1)] = f"[FAIL] {name_ex}:{output_id_ex}"
+
+                diffs[ptrn_ex.match(eo.stem).group(1)] = data_ex.flatten().size - np.count_nonzero(idx)
+
+    sorted_results = [val for key, val in sorted(results.items())]
+    sorted_diffs = [val for key, val in sorted(diffs.items())]
+    for r, d in zip(sorted_results, sorted_diffs):
+        print(r)
```
Comment (tsawada) [inline dlk/python/tools/test_debug.py]: don't print within unittests. Ret the tetst fail

Comment (tkng) [reply dlk/python/tools/test_debug.py]: This file is not for unittest. Maybe it's better to rename the filename?

Comment (tkng) [reply dlk/python/tools/test_debug.py]: This file is not for unittest. Maybe it's better to rename the filename?

Comment (nlpng) [reply dlk/python/tools/test_debug.py]: The file name is indeed too confusing, I will rename it.

File: dlk/python/tools/test_debug.py [hunk 3]
```diff
@@ -53,19 +53,34 @@ def main(debug_data_path, expected_data_path):
...
+            if name_ex == name_dbg and output_id_ex == output_id_dbg:
                 data_dbg = np.load(o)
                 data_ex = np.load(eo)
 
-                if np.allclose(data_ex.flatten() * data_dbg['scale'], data_dbg['data'], rtol=0.00001, atol=0.00001):
-                    print(f"[OK]   {name_ex}")
+                r_tol = 0.0001
+                a_tol = 0.0001
+
+                idx = np.isclose(data_ex.flatten() * data_dbg['scale'], data_dbg['data'], rtol=r_tol, atol=a_tol)
```
[the remark this answers is not in the document]
Comment (nlpng) [reply dlk/python/tools/test_debug.py]: How about `within_tolerance` (from @joelN123) for this variable?

[the remark this answers is not in the document]
Comment (tkng) [reply dlk/python/tools/test_debug.py]: sounds good!

**Comment by nlpng** [comment]
run test

**Comment by tkng** [comment]
- [x] readability approval
- [x] ownership approval

**Pull request merged**